client/aws/aws.go

```diff
@@ -92,109 +94,131 @@ type PriceRaw struct {
+			for _, rawOffer := range _offerList {
+				offer := &client.Offer{
+					ID: incrID,
+
+					InstanceSKU: instanceSKU,
+					// AWS only offer instance-wise product
+					OfferType:     client.OfferTypeInstance,
+					ChargeType:    client.ChargeType(chargeType),
+					ChargePayload: rawOffer.Term,
+				}
+				// an offer may have differnet charging dimension, say upfront fee and it relevant fee charged hourly.
+				for _, dimension := range rawOffer.Dimension {
 					USDFloat, err := strconv.ParseFloat(dimension.PricePerUnit[client.CurrencyUSD], 64)
 					if err != nil {
 						return nil, fmt.Errorf("Fail to parse the price to type FLOAT64, value: %v, [internal]: %v", dimension.PricePerUnit[client.CurrencyUSD], err)
 					}
-					price := &client.Offer{
-						ID:            string(rateCode),
-						InstanceID:    string(instanceID),
-						ChargeType:    client.ChargeType(chargeType),
-						ChargePayload: offer.Term,
-						Description:   dimension.Description,
-						Unit:          dimension.Unit,
-						USD:           USDFloat,
+					if dimension.Unit == "Quantity" {
+						offer.CommitmentUSD = USDFloat
+					} else {
+						offer.HourlyUSD = USDFloat
 					}
-					priceList = append(priceList, price)
 				}
+
+				incrID++
+				offerList = append(offerList, offer)
 			}
 		}
 	}
 
-	return priceList, nil
+	return offerList, nil
 }
 
-func extractInstance(rawData *rawJSON) ([]*client.Instance, error) {
-	byteProducts, err := json.Marshal(rawData.Product)
-	if err != nil {
-		return nil, fmt.Errorf("Fail to unmarshal the result, [internal]: %v", err)
-	}
-	productsDecoder := json.NewDecoder(bytes.NewReader(byteProducts))
-	var rawEntryList InstanceRecord
-	if err := productsDecoder.Decode(&rawEntryList); err != nil {
-		return nil, fmt.Errorf("Fail to decode the result, [internal]: %v", err)
+func fillInstancePayload(instanceRecord InstanceRecord, offerList []*client.Offer) {
+	// there are maybe many offer that are bind to the same instance (i.e. the same SKU).
+	offerMap := make(map[string][]*client.Offer)
+	for _, offer := range offerList {
+		if _, ok := offerMap[offer.InstanceSKU]; ok {
+			offerMap[offer.InstanceSKU] = append(offerMap[offer.InstanceSKU], offer)
+		} else {
+			offerMap[offer.InstanceSKU] = []*client.Offer{offer}
+		}
 	}
 
-	var instanceList []*client.Instance
-	for id, entry := range rawEntryList {
+	for instanceSKU, entry := range instanceRecord {
 		if entry.ProductFamily != "Database Instance" /* filter non-db instance */ ||
 			entry.Attributes.DeploymentOption != "Single-AZ" /* filter multi-region deployment */ {
 			continue
 		}
-		instance := &client.Instance{
-			ID:                 string(id),
-			ServiceCode:        entry.Attributes.ServiceCode,
-			Region:             entry.Attributes.Location,
+
+		engineType := entry.Attributes.DatabaseEngine.String()
+		if engineType == "UNKNOWN" {
```

Comment:
Do we have an enum or constant for this?

Reply:
A enum here actually function as a adpotor, mapping the engine named by AWS to the type stored in ours. 
e.g. AWS's `SQL Server` would be mapped to `SQLSERVER`.